src/librustc_trans/adt.rs

```diff
@@ -173,492 +172,261 @@ pub fn represent_type<'a, 'tcx>(cx: &CrateContext<'a, 'tcx>,
-        // let's recurse and find out
-        ty::TyAdt(def, substs) if def.is_struct() => {
-            for (j, field) in def.struct_variant().fields.iter().enumerate() {
-                let field_ty = monomorphize::field_ty(tcx, substs, field);
-                if let Some(mut fpath) = find_discr_field_candidate(tcx, field_ty, path.clone()) {
-                    fpath.push(j);
-                    return Some(fpath);
-                }
-            }
-            None
-        },
-
-        // Perhaps one of the upvars of this struct is non-zero
-        // Let's recurse and find out!
-        ty::TyClosure(_, ref substs) => {
-            for (j, &ty) in substs.upvar_tys.iter().enumerate() {
-                if let Some(mut fpath) = find_discr_field_candidate(tcx, ty, path.clone()) {
-                    fpath.push(j);
-                    return Some(fpath);
-                }
-            }
-            None
-        },
-
-        // Can we use one of the fields in this tuple?
-        ty::TyTuple(ref tys) => {
-            for (j, &ty) in tys.iter().enumerate() {
-                if let Some(mut fpath) = find_discr_field_candidate(tcx, ty, path.clone()) {
-                    fpath.push(j);
-                    return Some(fpath);
-                }
-            }
-            None
-        },
-
-        // Is this a fixed-size array of something non-zero
-        // with at least one element?
-        ty::TyArray(ety, d) if d > 0 => {
-            if let Some(mut vpath) = find_discr_field_candidate(tcx, ety, path) {
-                vpath.push(0);
-                Some(vpath)
-            } else {
-                None
-            }
-        },
-
-        // Anything else is not a pointer
-        _ => None
+fn mk_mach_int_from_layout<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                                      tm: layout::Integer,
+                                      signed: bool) -> Ty<'tcx> {
+    use rustc::ty::layout::Integer::*;
+    match (tm, signed) {
+        (I1, false) => tcx.types.u8,
+        (I8, false) => tcx.types.u8,
+        (I16, false) => tcx.types.u16,
+        (I32, false) => tcx.types.u32,
+        (I64, false) => tcx.types.u64,
+        (I1, true) => tcx.types.i8,
+        (I8, true) => tcx.types.i8,
+        (I16, true) => tcx.types.i16,
+        (I32, true) => tcx.types.i32,
+        (I64, true) => tcx.types.i64,
     }
 }
 
-impl<'tcx> Case<'tcx> {
-    fn is_zerolen<'a>(&self, cx: &CrateContext<'a, 'tcx>, scapegoat: Ty<'tcx>) -> bool {
-        mk_struct(cx, &self.tys, false, scapegoat).size == 0
-    }
-
-    fn find_ptr<'a>(&self, cx: &CrateContext<'a, 'tcx>) -> Option<DiscrField> {
-        for (i, &ty) in self.tys.iter().enumerate() {
-            if let Some(mut path) = find_discr_field_candidate(cx.tcx(), ty, vec![]) {
-                path.push(i);
-                return Some(path);
-            }
-        }
-        None
+fn convert_layout_struct<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
```

Comment:
This function should be dead by the time I'm done as well. I'm not sure what action and/or problem you see with it though.  Can you explain further?